app/services/transaction_service/transaction.rb

```diff
@@ -369,27 +367,23 @@ def checkout_details(model)
   end
 
   def calculate_commission(total_price, commission_from_seller, minimum_commission)
-    if commission_from_seller.blank? || commission_from_seller == 0
-      Money.new(0, minimum_commission.currency)
-    else
-      commission_by_percentage = total_price * (commission_from_seller / 100.0)
-      (commission_by_percentage > minimum_commission) ? commission_by_percentage : minimum_commission
-    end
+    [(total_price * (commission_from_seller / 100.0) unless commission_from_seller.nil?),
+     (minimum_commission unless minimum_commission.nil? || minimum_commission.zero? ),
+     Money.new(0, total_price.currency)]
+      .compact
+      .max
   end
 
-  def calculate_commission_to_admin(commission_total, fee_total)
-    commission_total - fee_total #we charge from admin, only the sum after paypal fees
+  def calculate_commission_to_admin(commission_total, payment_total, fee_total)
+    # Ensure we never charge more than what the seller received after payment processing fee
+    [commission_total, payment_total - fee_total].min
   end
 
-  def paypal_account_verified?(community_id:, person_id:nil)
-    acc = paypal_accounts_api.get(
-      community_id: community_id,
-      person_id: person_id
-    ).maybe
+  def paypal_account_verified?(community_id:, person_id: nil, settings: Maybe(nil))
```

Comment:
Conceptually it's wrong to reach inside transaction service in controllers to do a purely functional data validation. Equally wrong would be to call from a service implementation to PaypalHelper controller utility. This logic is part of a "PayPal service client", i.e. something that is implemented at the side that calls out to paypal service. To share it would mean to implement a client package and introduce it as a dependency to anyone who needs to call out to paypal service. But that's clearly overkill now so instead we have this "intentional copy & paste". One way to get around this in some scenarios is to introduce calculated enum fields to the service resource representation but that's always a trade-off because then you end up calculating and sharing the data to everyone whether or not they need it.